### DISCUSSION TITLE
Rewrite in C++?

### DISCUSSION THREAD

**Discussion by jrosdahl** [ccache/ccache#415] [Ideas]
@pavsa wrote this in https://github.com/ccache/ccache/issues/404#issuecomment-489412432:

> Just a crazy question, wouldn’t it be more productive / less error prone to rewrite it to C++? No crazy template magic, just “C with classes” approach. There’s lot of allocate/free on error patterns throughout the code, which would be unnecessary with RAII. Also a lot of built in hash tables, vectors, so on. Testing would be also easier.

Moving the discussion to a separate issue.

**Comment by jrosdahl** [comment]
Sure, rewriting in C++ would be feasible, but I'm not sure if the time and effort to do so would be worth it.

There is also the issue that a dependency on C++ most likely will affect portability. Which may be OK, but it needs to be taken into account.

> Testing would be also easier.

@pavsa: In what way? Do you refer to better test frameworks, easier dependency injection, more natural use of class interfaces or something else?

**Comment by jrosdahl** [comment]
@pavsa: Any input on my question above?

**Comment by pavsa** [comment]
> 
> 
> Sure, rewriting in C++ would be feasible, but I'm not sure if the time and effort to do so would be worth it.
> 
> There is also the issue that a dependency on C++ most likely will affect portability. Which may be OK, but it needs to be taken into account.
> 
> > Testing would be also easier.
> 
> @pavsa: In what way? Do you refer to better test frameworks, easier dependency injection, more natural use of class interfaces or something else?

All above - standard test driven development. 
Everything that can be done in C++ can of course be done in C, but C code often ends up being larger and more error prone. You sometimes have to reinvent a wheel in C, while C++ offers you quite a lot in STL.

**Comment by jrosdahl** [comment]
> Everything that can be done in C++ can of course be done in C, but C code often ends up being larger and more error prone. You sometimes have to reinvent a wheel in C, while C++ offers you quite a lot in STL.

I'm aware. I've used C++ professionally for 20 years now (and on and off 6 years before that).

*If* we switch to C++, it will likely not be a full big bang rewrite but a gradual introduction.

**Comment by jrosdahl** [comment]
Closing this discussion now since it seems to have ended.

Related: I'm currently working on some new internal APIs, and using C feels increasingly silly... I think that a gradual introduction very well might happen in the future.

**Comment by pavsa** [comment]
Looking forward to that 👍
